### PR TITLE
Placeholder component

### PR DESCRIPTION
**Pull request by azu-amdocs**

### CODE REVIEW

File: src/Card/Card.jsx [hunk 1]
```diff
@@ -6,6 +6,7 @@ import CardHeader from './CardHeader';
 import buildThemeProperties from '../Utils';
 
 import style from '../Layout/LayoutStyle.scss';
+import PlaceholderTextComponent from '../../dev/src/components/PlaceholderComponent';
```
Comment (Tri4Ty) [inline src/Card/Card.jsx:9]: why is this in the '/dev/src' directory? Production code should not be in '/dev', it should be under '/src/'

File: src/Card/CardHeader.jsx [hunk 2]
```diff
@@ -26,18 +27,26 @@ export default class CardHeader extends Component {
...
+          {title ? <div className={themeStyles.titleWrapper}><span className={themeStyles.title} title={title}>{title}</span></div> : null }
+        </div>
+      );
+    }
     return (
-      <div className={this.createHeaderClassName()}>
-        {actions ? <CardActions {...this.props} /> : null }
-        {title ? <div className={themeStyles.titleWrapper}><span className={themeStyles.title} title={title}>{title}</span></div> : null }
-      </div>
+      <PlaceholderTextComponent {...this.props} />
```
Comment (Tri4Ty) [inline src/Card/CardHeader.jsx:39]: The text prop of PlaceHolderTextComponent is a string... the render method of PlaceHolderTextComponent simply replaces any occurrence of a regex with a span with the appropriate css style. so in the end we might have:

`some title <span className={classname}></span>`

I was thinking we'd have something like

`<span title="some calculated title string">some title <span className={classname}></span></span>`

this way, when people hover on it (or screen reader), we have an actual title property

File: src/Card/Card.jsx (hunk 1 above)
Comment (azu-amdocs) [reply src/Card/Card.jsx:9]: Ok.

File: src/Card/Card.jsx [hunk 3]
```diff
@@ -82,7 +82,10 @@ export default class Card extends Component {
 
 Card.propTypes = {
   configId: PropTypes.string.isRequired,
-  title: PropTypes.string,
+  title: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.instanceOf(Object),
```
Comment (Tri4Ty) [inline src/Card/Card.jsx:87]: we should be explicit here ... instead of instanceOf(Object), this should be PropTypes.instanceOf(PlaceholderComponent)

File: src/Card/CardHeader.jsx [hunk 4]
```diff
@@ -6,6 +6,7 @@ import CardActions from './CardActions';
 import buildThemeProperties from '../Utils';
 
 import style from '../Layout/LayoutStyle.scss';
+import PlaceholderTextComponent from '../PlaceholderComponent';
```
Comment (Tri4Ty) [inline src/Card/CardHeader.jsx:9]: no need to import this... the PlaceholderComponent element will be passed in as a prop (it's up to the parent/containing component to create the PlaceholderComponent and pass it in as the title prop

File: src/Card/CardHeader.jsx [hunk 5]
```diff
@@ -26,18 +27,26 @@ export default class CardHeader extends Component {
       compose: 'Merge',
     });
 
+    if (typeof title === 'string') {
```
Comment (Tri4Ty) [inline src/Card/CardHeader.jsx:30]: can clean up/simplify this if condition by using "inline if" statement (https://reactjs.org/docs/conditional-rendering.html#inline-if-with-logical--operator) ... instead of wrapping and duplicating the return, you can simply inline if around the title portion of the return. You can already see the use of inline condition ... you just need to add checks

{ title && typeof title === 'string' ? [render what's currently there] : [the title is a Placeholder, so just render it] }

File: src/Card/CardHeader.jsx (hunk 2 above)
Comment (Tri4Ty) [reply src/Card/CardHeader.jsx:39]: totally don't need the <Placeholder> tag here... it's already passed to you... see previous comment around inline if

File: src/Card/CardHeader.jsx [hunk 6]
```diff
@@ -26,18 +27,26 @@ export default class CardHeader extends Component {
...
     );
   }
 }
 
 CardHeader.propTypes = {
   cardId: PropTypes.string.isRequired,
-  title: PropTypes.string,
+  title: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.instanceOf(Object),
```
Comment (Tri4Ty) [inline src/Card/CardHeader.jsx:48]: we should be explicit here ... instead of instanceOf(Object), this should be PropTypes.instanceOf(PlaceholderComponent)

File: src/PlaceholderComponent.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import reactStringReplace from 'react-string-replace';
+
+
+function PlaceholderTextComponent({ title }) {
```
Comment (Tri4Ty) [inline src/PlaceholderComponent.js:6]: What happened to the design on the team design share? there were 3 props passed in to this component:

- text 
- regex
- cssClass

why was that changed? Looks like you embeded those props inside the title prop

File: src/PlaceholderComponent.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      {reactStringReplace(title.text, title.regex, (match, i) => (
+        <span key={i} className={title.cssClass} />
+      ))
+    }
+    </React.Fragment>
+  );
+}
+
+PlaceholderTextComponent.propTypes = {
+  title: PropTypes.oneOfType([
```
Comment (Tri4Ty) [inline src/PlaceholderComponent.js:18]: this is confusing... your render function uses a single title prop but always assumes its an object with nested text, regex and cssClass attributes... this seems inconsistent with that.
In saying this, as discussed at the top of this class, we originally agreed on 3 props being passed in

File: src/PlaceholderComponent.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    PropTypes.string,
+    PropTypes.instanceOf(Object),
+  ]),
+};
+
+PlaceholderTextComponent.defaultProps = {
```
Comment (Tri4Ty) [inline src/PlaceholderComponent.js:24]: I thought there was a default value for regex? The other props don't have default values so can be left out

**Review by Tri4Ty — CHANGES_REQUESTED** [review]
Several comments in the review... seems we've deviated from the original design... lets sync up on a call to discuss all this

**Review by Tri4Ty — APPROVED** [review]
